### PR TITLE
Introduce frame elements

### PR DESCRIPTION
**Pull request by pinussilvestrus**
This pull request introduces basic frame elements support

* [x] Allow creating elements with _frame_ character (cf. doc 5e9bca67ba825de10edd54b9189309d22d7011e3)
* [x] Frame elements are only interactable on border (cf. ef732eee83c3e3fee59e6266c2f433fbb95848aa)
* [x] It is not allowed to drop elements on frame borders
* [x] Add visualization (hover, select, drag.hover,  cf. 13cc9f8d7c5a5a51c775fafbb38db509c567f048, 9d81e48239c7dc09a16715e3117c23d96ff562a1)

Relates to bpmn-io/bpmn-js#959
Relates to bpmn-io/bpmn-js#960

**Frame Interaction**

Apr-30-2019 08-48-26

**Prevent dropping on borders**

Apr-30-2019 08-57-39

**Note**

We decided to _not_ include the second (inner) outline to frame element borders, since it could lead to confusions (other tools does not offer it at all). Furthermore, right now we only think about rectangles as possible frame elements. The inner square outline would look weird with circle elements (e.g. when creating custom elements).

Let's try out whether only one outline is enough for the user to determine border-only interaction. When we find out that this is confusing, we could add it in a second iteration. I moved the code to a standalone extension (cf. [`diagram-js-frame-outline`)](https://github.com/pinussilvestrus/diagram-js-frame-outline)

### CODE REVIEW

**Comment by pinussilvestrus** [comment]
We decided to _not_ include the second (inner) outline to frame element borders, since it could lead to confusions (other tools does not offer it at all). Furthermore, right now we only think about rectangles as possible frame elements. The inner square outline would look weird with circle elements (e.g. when creating custom elements).

Let's try out whether only one outline is enough for the user. When we find out that this is confusing, we could add it in a second iteration. I moved the code to a standalone extension (cf. [`diagram-js-frame-outline`)](https://github.com/pinussilvestrus/diagram-js-frame-outline)

**Comment by pinussilvestrus** [comment]
Note for the reviewer: To test in a modeling manner, please refer to the `bpmn-js` [implementation example branch](https://github.com/bpmn-io/bpmn-js/tree/group-frame-support).

File: lib/model/index.js [hunk 1]
```diff
@@ -105,6 +105,16 @@ export function Base() {
 export function Shape() {
   Base.call(this);
 
+  /**
+   * Indicates frame shapes
+   *
+   * @name Shape#frameOnly
+   * @type Boolean
+   */
+  Object.defineProperty(this, 'frameOnly', {
```
Comment (nikku) [inline lib/model/index.js]: If we `Object#defineProperty` here the property won't be enumerable (and JSON#stringifiable). I'd suggest we do not define the property but keep the JSDoc to indicate that shape property.

File: lib/features/create/Create.js [hunk 2]
```diff
@@ -3,7 +3,8 @@ var LOW_PRIORITY = 750;
 var MARKER_OK = 'drop-ok',
     MARKER_NOT_OK = 'drop-not-ok',
     MARKER_ATTACH = 'attach-ok',
-    MARKER_NEW_PARENT = 'new-parent';
+    MARKER_NEW_PARENT = 'new-parent',
+    MARKER_FRAME_NOT_OK = 'frame-drop-not-ok';
```
Comment (nikku) [inline lib/features/create/Create.js]: Why do we need a separate class here? Couldn't we add a class such as `djs-frame` to `djs-shape` elements?

This way users could define specialized CSS declarations like `djs-frame.drop-not-ok` (or the like) and we'd not need to touch the `Create` and `Move` features at all.

File: lib/features/interaction-events/InteractionEvents.js [hunk 3]
```diff
@@ -45,10 +49,20 @@ var LOW_PRIORITY = 500;
...
 
-  var HIT_STYLE = styles.cls('djs-hit', [ 'no-fill', 'no-border' ], {
-    stroke: 'white',
-    strokeWidth: 15
-  });
+  function createHitStyle(element) {
+
+    var traits = [ 'no-fill', 'no-border' ];
+
+    if (isFrameElement(element)) {
```
Comment (nikku) [inline lib/features/interaction-events/InteractionEvents.js]: Even this could be handled via CSS, if a frame had the `djs-shape.djs-frame` classes.

**Comment by pinussilvestrus** [comment]
Thanks for reviewing! I'll check whether we can rid of the extra stuff by adding the `.djs-frame` class.

File: lib/core/GraphicsFactory.js [hunk 4]
```diff
@@ -92,10 +92,12 @@ GraphicsFactory.prototype._clear = function(gfx) {
  *   <g class="djs-children"></g>
  * </g>
  *
- * @param {Object} parent
  * @param {String} type the type of the element, i.e. shape | connection
+ * @param {Snap<SVGElement>} [childrenGfx]
```
Comment (nikku) [inline lib/core/GraphicsFactory.js]: We're not returning Snap<SVGElement> here, instead only raw SVG elements.

Comment (philippfromme) [reply lib/core/GraphicsFactory.js]: Good old times.

Comment (nikku) [reply lib/core/GraphicsFactory.js]: :slightly_smiling_face: :old_key:

Comment (pinussilvestrus) [reply lib/core/GraphicsFactory.js]: updated the return docs.

**Comment by pinussilvestrus** [comment]
Update: I'll add some test coverage for the `connect-not-ok` markers.

**Pull request merged**